L1Trigger/L1TTrackMatch/plugins/L1TkMuonProducer.cc

```diff
@@ -0,0 +1,712 @@
+  void build_tkMuons_from_idxs(TkMuonCollection& tkMuons,
+                               const std::vector<int>& matches,
+                               const edm::Handle<L1TTTrackCollectionType>& l1tksH,
+                               const edm::Handle<RegionalMuonCandBxCollection>& muonH,
+                               int detector) const;
+
+  // as above, but for the TkMu that were built from a EMTFCollection - do not produce a valid muon ref
+  void build_tkMuons_from_idxs(TkMuonCollection& tkMuons,
+                               const std::vector<int>& matches,
+                               const edm::Handle<L1TTTrackCollectionType>& l1tksH,
+                               int detector) const;
+
+  // dump and convert tracks to the format needed for the MAnTra correlator
+  std::vector<L1TkMuMantraDF::track_df> product_to_trkvec(const L1TTTrackCollectionType& l1tks);  // tracks
+  std::vector<L1TkMuMantraDF::muon_df> product_to_muvec(
+      const RegionalMuonCandBxCollection& l1mtfs);  // regional muon finder
+  std::vector<L1TkMuMantraDF::muon_df> product_to_muvec(
+      const EMTFTrackCollection& l1mus);  // endcap muon finder - eventually to be moved to regional candidate
+
+  float ETAMIN_;
+  float ETAMAX_;
+  float ZMAX_;  // |z_track| < ZMAX in cm
+  float CHI2MAX_;
+  float PTMINTRA_;
+  float DRmax_;
+  int nStubsmin_;  // minimum number of stubs
+  bool correctGMTPropForTkZ_;
+  bool use5ParameterFit_;
+  bool useTPMatchWindows_;
+
+  AlgoType bmtfMatchAlgoVersion_;
+  AlgoType omtfMatchAlgoVersion_;
+  AlgoType emtfMatchAlgoVersion_;
+
+  std::unique_ptr<L1TkMuCorrDynamicWindows> dwcorr_;
+
+  std::unique_ptr<L1TkMuMantra> mantracorr_barr_;
+  std::unique_ptr<L1TkMuMantra> mantracorr_ovrl_;
+  std::unique_ptr<L1TkMuMantra> mantracorr_endc_;
+  int mantra_n_trk_par_;
+
+  const edm::EDGetTokenT<RegionalMuonCandBxCollection> bmtfToken;
+  const edm::EDGetTokenT<RegionalMuonCandBxCollection> omtfToken;
+  const edm::EDGetTokenT<RegionalMuonCandBxCollection> emtfToken;
+  const edm::EDGetTokenT<EMTFTrackCollection>
+      emtfTCToken;  // the track collection, directly from the EMTF and not formatted by GT
+  const edm::EDGetTokenT<std::vector<TTTrack<Ref_Phase2TrackerDigi_> > > trackToken;
+};
+
+L1TkMuonProducer::L1TkMuonProducer(const edm::ParameterSet& iConfig)
+    : bmtfToken(consumes<RegionalMuonCandBxCollection>(iConfig.getParameter<edm::InputTag>("L1BMTFInputTag"))),
+      omtfToken(consumes<RegionalMuonCandBxCollection>(iConfig.getParameter<edm::InputTag>("L1OMTFInputTag"))),
+      emtfToken(consumes<RegionalMuonCandBxCollection>(iConfig.getParameter<edm::InputTag>("L1EMTFInputTag"))),
+      emtfTCToken(consumes<EMTFTrackCollection>(iConfig.getParameter<edm::InputTag>("L1EMTFTrackCollectionInputTag"))),
+      trackToken(consumes<std::vector<TTTrack<Ref_Phase2TrackerDigi_> > >(
+          iConfig.getParameter<edm::InputTag>("L1TrackInputTag"))) {
+  ETAMIN_ = (float)iConfig.getParameter<double>("ETAMIN");
```

Comment:
these can all be initialized in the constructor initializer list

Reply:
done